### PR TITLE
Use `-Z rustdoc-map` instead of rewriting it

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/docs.rs/issues/1177, closes https://github.com/rust-lang/docs.rs/issues/1165. Fixes https://github.com/rust-lang/docs.rs/issues/1166.

### CODE REVIEW

**Comment by Nemo157** [comment]
A crate to test #1166 with is `futures-util-preview 0.2.2`, it depends on `futures-core-preview` which has the lib name `futures_core`.

**Comment by Nemo157** [comment]
Tested with this little patch applied on top:

```patch
diff --git src/docbuilder/rustwide_builder.rs src/docbuilder/rustwide_builder.rs
index 83592dc..1a80aea 100644
--- src/docbuilder/rustwide_builder.rs
+++ src/docbuilder/rustwide_builder.rs
@@ -579,6 +579,9 @@ impl RustwideBuilder {

         // Add docs.rs specific arguments
         let mut cargo_args = Vec::new();
+        cargo_args.push("--verbose".into());
+        cargo_args.push(r#"--config=doc.extern-map.registries.crates-io="https://docs.rs""#.into());
+        cargo_args.push("-Zunstable-options".into());
         if let Some(cpu_limit) = self.config.build_cpu_limit {
             cargo_args.push(format!("-j{}", cpu_limit));
         }
```

On `master` I see us passing:

```
"--extern-html-root-url" "futures_core_preview=https://docs.rs/futures-core-preview/0.2.3"
```

And on this branch I see `cargo` passing:

```
--extern-html-root-url 'futures_core=https://docs.rs/futures-core-preview/0.2.3/'
```

So this definitely fixes #1166

**Comment by jyn514** [comment]
Thanks, that definitely looks like the right approach since it means the hyperlinks will work even when built with old versions of nightly :)

This is ready for re-review.

File: src/docbuilder/rustwide_builder.rs [hunk 1]
```diff
@@ -589,6 +579,12 @@ impl RustwideBuilder {
 
         // Add docs.rs specific arguments
         let mut cargo_args = Vec::new();
+        // Show the command line Rustdoc is invoked with
+        cargo_args.push("--verbose".into());
+        // We know that `metadata` unconditionally passes `-Z rustdoc-map`.
+        // Don't copy paste this, since that fact is not stable and may change in the future.
+        // Normally we would need -Z unstable-options, but we currently pass that too.
```
Comment (Nemo157) [inline src/docbuilder/rustwide_builder.rs:586]: We didn't in my testing, I think it's only when the crate has `package.metadata.docs.rs.rustc-args` set that we do.

**Comment by pietroalbini** [comment]
Random note on this: it'd be nice if `rustdoc-map` allowed using the semver pattern instead of the version, since docs.rs would then redirect to the latest semver-compatible docs automatically:

```
https://docs.rs/rustwide/^0.10/...
```

**Comment by jyn514** [comment]
That would be nice, yeah - it needs rustdoc changes first, though.
https://github.com/rust-lang/docs.rs/issues/1055

**Pull request merged**